proof/crefine/AARCH64/Tcb_C.thy

```diff
@@ -4505,6 +4505,104 @@ lemma decodeSetTLSBase_ccorres:
   apply (auto simp: unat_eq_0 le_max_word_ucast_id)+
   done
 
+lemma scast_seL4_TCBFlag_fpuDisabled[simp]:
+  "scast seL4_TCBFlag_fpuDisabled = tcbFlagToWord (ArchFlag FpuDisabled)"
+  by (clarsimp simp: seL4_TCBFlag_fpuDisabled_def tcbFlagToWord_def archTcbFlagToWord_def)
+
+lemma invokeTCB_SetFlags_ccorres:
+  notes hoare_weak_lift_imp [wp]
+  shows
+  "ccorres (cintr \<currency> (\<lambda>rv rv'. rv = [])) (liftxf errstate id (K ()) ret__unsigned_long_')
+   (invs' and tcb_at' tcb)
+   ({s. thread_' s = tcb_ptr_to_ctcb_ptr tcb}
+         \<inter> {s. clear_' s = clears}
+         \<inter> {s. set_' s = sets}) []
+   (invokeTCB (SetFlags tcb clears sets))
+   (Call invokeSetFlags_'proc)"
+  apply (cinit lift: thread_' clear_' set_' simp: setFlags_def postSetFlags_def)
+   apply (simp add: liftE_def bind_assoc)
+   apply (rule ccorres_pre_threadGet, rename_tac flags)
+   \<comment> \<open>split off flags_' updates and threadSet\<close>
+   apply (rule ccorres_rhs_assoc2, rule ccorres_rhs_assoc2, rule ccorres_rhs_assoc2)
+   apply (rule ccorres_split_nothrow)
+       apply (rule_tac P="\<lambda>tcb. tcbFlags tcb = flags" in threadSet_ccorres_lemma2)
+        apply vcg
+       apply (clarsimp simp: tcb_at_h_t_valid[OF obj_tcb_at'])
+       apply (erule rf_sr_tcb_update_no_queue2[rotated],
+              (simp add: typ_heap_simps')+, simp_all?)[1]
+        apply (rule ball_tcb_cte_casesI, simp+)
+       apply (clarsimp simp: ctcb_relation_def)
+      apply ceqv
+     apply (rule ccorres_split_nothrow_novcg_dc)
+        apply (rule_tac R=\<top> and R'="{s. flags_' s = flags && ~~ clears || sets}" in ccorres_when_strong)
+         apply (clarsimp simp: isFlagSet_def word_bw_comms)
+        apply (ctac add: fpuRelease_ccorres)
+       apply (rule ccorres_return_CE[folded return_returnOk]; simp)
+      apply (wpsimp simp: guard_is_UNIV_def)+
+   apply vcg
+  apply (clarsimp simp: obj_at'_def typ_heap_simps' ctcb_relation_def)
+  done
+
+lemma decodeSetFlags_ccorres:
+  "ccorres (intr_and_se_rel \<currency> dc) (liftxf errstate id (K ()) ret__unsigned_long_')
+       (invs' and sch_act_simple
+              and (\<lambda>s. ksCurThread s = thread) and ct_active' and K (isThreadCap cp)
+              and valid_cap' cp and (\<lambda>s. \<forall>x \<in> zobj_refs' cp. ex_nonz_cap_to' x s)
+              and sysargs_rel args buffer)
+       (UNIV
```

Comment:
As this is a new lemma, we don't want the UNIV, and I would prefer if you used the modern variable binding style (with acutes and {| |} brackets).
also the >>= is quite deeply indented below